### DISCUSSION TITLE
Appender API with nested types

### DISCUSSION THREAD

**Discussion by loicalleyne** [marcboeker/go-duckdb#309] [Q&A]
Today I saw a post on LinkedIn by an engineer at ShareChat [link](https://www.linkedin.com/feed/update/urn:li:activity:7258424841538555904?updateEntityUrn=urn%3Ali%3Afs_updateV2%3A%28urn%3Ali%3Aactivity%3A7258424841538555904%2CFEED_DETAIL%2CEMPTY%2CDEFAULT%2Cfalse%29) where they managed to get 20K inserts/sec using a query batcher but estimate that appender could be even faster. The table schema they had was flat, and usually I have to deal with nested data; I've seen it said that there is an 'append data chunk' way of using appender but in go-duckdb that function is not public. Is there a way to use appender with nested data using the existing AppendRow function?
Or is it easier/faster to use the Arrow RegisterView and insert a select from the view?

**Comment by taniabogatsch** [comment]
Hi @loicalleyne!

> I've seen it said that there is an 'append data chunk' way of using appender but in go-duckdb that function is not public

Indeed, when using the C API, we recommend using the [`duckdb_append_data_chunk`](https://github.com/duckdb/duckdb/blob/1fe66c2ca4d6ae5aa88a46df6b9d4406e19ed35f/src/include/duckdb.h#L3771) functions. That way, the client/user can pass data chunks to the appender, which stores them in a local `ColumnDataCollection` and eventually batch inserts them.

Internally, go-duckdb also uses `duckdb_append_data_chunk`. The `Appender` also uses these data chunks. When you call `AppendRow`, it fills the current data chunk (row-wise). Eventually, go-duckdb flushes all data chunks via `duckdb_append_data_chunk`.

Another way to fill these data chunks internally would be to expose an `AppendColumn` function. go-duckdb does not implement that. We've only discussed it in a few PRs.

So, to summarize, the `Appender` and `AppendRow` internally already use the recommended C API functions to append chunks. However, go-duckdb still has room for improvement in how it creates/fills these data chunks.

> Is there a way to use appender with nested data using the existing AppendRow function?

Yes, this is possible. Look at [this test case](https://github.com/marcboeker/go-duckdb/blob/e1b55d8ee6dbb1b9f89272ca9c50d2f7af450cf6/appender_test.go#L255), which appends many different nested types.

> Or is it easier/faster to use the Arrow RegisterView and insert a select from the view?

I cannot say, I haven't benchmarked this.

Hope this helps - :)

**Comment by loicalleyne** [comment]
@taniabogatsch I've been busy working on [bodkin](https://github.com/loicalleyne/bodkin) a helper library for Arrow so I haven't tried the Appender yet; that being said I just tested inserting 836855 rows of data converted into Arrow Records into duckdb with 
RegisterView and a QueryContext `INSERT into T1 BY NAME (SELECT * FROM 'arrowview')` and it completed in 7.8 seconds or ~107K/rows per second. The data's schema is 46 fields with nested structs and lists.

**Comment by loicalleyne** [comment]
I finally had a chance to try the Appender API, and wasn't able to get an insertion rate over 27K rows/sec (it's actually quite variable and can sometimes be much lower, I might have to the same test on a more powerful machine than my laptop). For the same shape of data ADBC has much better performance, the insertion rate caps out at ~75K rows/sec.

**Comment by taniabogatsch** [comment]
That's interesting! There's a current PR with an Appender improvement. I haven't benchmarked/profiled the (Go) Appender too extensively. It would be interesting to see how close it could get to your ADBC timings.

**Reply by loicalleyne** [reply]
It's probably pertinent to mention that the performance difference mentioned above isn't even counting the format conversion from the source data to the shape needed for the API. 

The project I'm working on is (1) reading protobuf messages from Kafka, (2) deserializing them and converting the format (3) using the API to append to the duckdb table. 

The end goal would be to run all these processes concurrently but to make identifying bottlenecks easier :
- steps (1),(2) and (3) are executed in series
- make a buffered channel for each step's output type which matches the input type of the following step
    - (1) chan []byte
    - (2) chan arrow.Record or chan driver.Value
- steps (1) and (2) output to their output channel and close the channel when done
- steps (2), and (3) range over the input channel

Observations:
- Reading rate from Kafka is network-bound (saturating network connection with more goroutines increases rate)
- Protobuf deserialization is CPU-bound (more goroutines and more powerful CPU increases rate)
- Format conversion is CPU-bound 
    - converting protobuf data to arrow (using [bufarrow](https://github.com/loicalleyne/bufarrow) which converts the proto schema to arrow and generates a builder with all the field build functions) is much faster than deserializing, converting to a map, type asserting to `driver.Value` and appending to []driver.Value for appender.AppendRow([]drive.Value).
    
Here are the logs from my testing comparing using Appender API vs. ADBC with writing directly to parquet for comparison.
ADBC has similar insertion rates on my laptop vs. a very beefy cloud VM which makes me think there's perhaps room for optimization.
```
        Appender API workflow
2025/01/20 22:00:27 kafka read start with 5 readers
2025/01/20 22:00:30 read 1843200 kafka records in 3.247321secs @ 567606.267080 messages/sec
2025/01/20 22:00:30 deserialize to map[string]any with 15 goroutines start
2025/01/20 22:00:44 []byte deserialize to map[string]any done - 13.253302117 @ 139074.773317 messages/sec
2025/01/20 22:00:44 map to driver.Value start
2025/01/20 22:00:46 map to driver.Value done - 2.224372573 secs @ 828638.061196 messages/sec
2025/01/20 22:00:46 appender start
2025/01/20 22:02:15 duck insert 1843200 records in 89.252323073 secs @ 20651.563108 rows/sec
2025/01/20 22:02:15 duck Appender total duration: 107.977517 secs

        ABDC workflow
2025/01/20 21:50:54 kafka read start with 5 readers
2025/01/20 21:50:57 read 1843201 kafka records in 3.046837 secs @ 604955.621585 messages/sec
2025/01/20 21:50:57 deserialize []byte to proto, convert to arrow records with 15 goroutines start
2025/01/20 21:51:01 deserialize to arrow done - 1843201 records in 3.855908 secs @ 478019.954702 messages/sec
2025/01/20 21:51:01 ADBC IngestCreateAppend start
2025/01/20 21:51:03 rows ingested: 122880 - 67553.600880 rows/sec
2025/01/20 21:51:05 rows ingested: 122880 - 73580.838323 rows/sec
2025/01/20 21:51:06 rows ingested: 122880 - 73099.345628 rows/sec
2025/01/20 21:51:08 rows ingested: 122880 - 74068.716094 rows/sec
2025/01/20 21:51:10 rows ingested: 122880 - 70905.943451 rows/sec
2025/01/20 21:51:11 rows ingested: 122880 - 73979.530403 rows/sec
2025/01/20 21:51:13 rows ingested: 122880 - 73055.885850 rows/sec
2025/01/20 21:51:15 rows ingested: 122880 - 72710.059172 rows/sec
2025/01/20 21:51:15 rows ingested: 995 - 17456.140351 rows/sec
2025/01/20 21:51:15 rows ingested: 1027 - 26333.333333 rows/sec
2025/01/20 21:51:15 rows ingested: 283 - 9758.620690 rows/sec
2025/01/20 21:51:17 rows ingested: 120787 - 66257.268239 rows/sec
2025/01/20 21:51:18 rows ingested: 1016 - 644.670051 rows/sec
2025/01/20 21:51:18 rows ingested: 383 - 6719.298246 rows/sec
2025/01/20 21:51:18 rows ingested: 1435 - 29285.714286 rows/sec
2025/01/20 21:51:20 rows ingested: 122152 - 64699.152542 rows/sec
2025/01/20 21:51:22 rows ingested: 122055 - 53136.700044 rows/sec
2025/01/20 21:51:24 rows ingested: 122140 - 63880.753138 rows/sec
2025/01/20 21:51:27 rows ingested: 121032 - 52622.608696 rows/sec
2025/01/20 21:51:27 rows ingested: 1762 - 25171.428571 rows/sec
2025/01/20 21:51:27 rows ingested: 1505 - 32717.391304 rows/sec
2025/01/20 21:51:29 rows ingested: 121391 - 65687.770563 rows/sec
2025/01/20 21:51:31 rows ingested: 122198 - 48742.720383 rows/sec
2025/01/20 21:51:31 duck ADBC insert 1843201 records in 30.107398447 secs @ 61220.865722 rows/sec
2025/01/20 21:51:31 ADBC total duration: 37.010298 secs

        Write to parquet workflow
2025/01/20 21:52:42 kafka read start with 5 readers
2025/01/20 21:52:45 read 1843200 kafka records in 2.999530 secs - 614496.270468 messages/sec
2025/01/20 21:52:45 deserialize []byte to proto, convert to arrow records with 15 goroutines start
2025/01/20 21:52:49 deserialize to arrow done - 1843200 records in 3.749819 secs @ 491543.755589 messages/sec
2025/01/20 21:52:49 write parquet start with 30 writers
2025/01/20 21:52:56 parquet write with 30 writers - time: 1843202 in 7.881999 secs @ 233849.553896 records/sec
2025/01/20 21:52:56 parquet write total duration: 14.631493
```

**Reply by loicalleyne** [reply]
Update: 
I hadn't realized that it is possible to use multiple appenders concurrently on separate connections, the above benchmarks are using a single thread. 
I was able to get up to ~63K records/sec insertion rate using 15 concurrent connections.
On the other hand, while refactoring the appender code I realized my prior attempts to use the ADBC API concurrently were incorrect, and it _is_ possible to use ADBC concurrently with multiple connections. Based on this new implementation, when making sure to batch arrow records in batches of 122880 or greater (to align with duckdb storage row group size), ABDC is much, much faster than Appender. See below:
```
ABDC workflow
2025/01/23 23:39:27 kafka read start with 8 readers
2025/01/23 23:39:41 read 15728642 kafka records in 14.385530 secs @ 1093365.498477 messages/sec
2025/01/23 23:39:41 deserialize []byte to proto, convert to arrow records with 32 goroutines start
2025/01/23 23:40:04 deserialize to arrow done - 15728642 records in 22.283532 secs @ 705841.509812 messages/sec
2025/01/23 23:40:04 ADBC IngestCreateAppend start with 32 connections
2025/01/23 23:40:25 duck ADBC insert 15728642 records in 21.145649535 secs @ 743824.007783 rows/sec
```

**Comment by mlafeldt** [comment]
👋 Thanks for starting this discussion!

This repository has moved to https://github.com/duckdb/duckdb-go

If this topic is still relevant, please continue the conversation at the new location. I'm closing this now as part of the migration.

Please see https://github.com/marcboeker/go-duckdb/issues/565 for more details.
